### DISCUSSION TITLE
A few questions on the DG Capabilities in MOOSE

### DISCUSSION THREAD

**Discussion by pzimbrod** [idaholab/moose#17383] [Q&A General] [answered]
Hi all,

I'm currently setting up MOOSE to help me model the complex physics within metal additive manufacturing (Powder Bed Fusion, to be precise). As there are a lot of length scales involved, the capabilities of MOOSE are really helping me out and I'm loving the framework and its structure, thanks for the extensive work so far!

As I'm also dealing with rather complex advection-diffusion equations and I have a background coming from FV based methods, I was looking into the DG capabilities of MOOSE. My understanding is that it provides a useful formulation for such problems as CG-FEM falls behind on solving hyperbolic PDEs (see e.g. <sup>1</sup>,  Theorem 5.3 p.222) and FV has obvious limitations regarding p-refinement (which to my understanding is especially nice when solving in parallel to reduce MPI communication).
I've looked into the documentation of the DG Kernels along with the test cases and came across a few questions regarding the implementation:

1. First of all, I'm wondering why you need to specify the DG Kernels separately as an addition to a regular kernel, instead of replacing it - see e.g.:

[...]
```moose
[Variables]
  [./u]
    order = FIRST
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./time_u]
    type = TimeDerivative
    variable = u
  [../]

  [./adv_u]
    implicit = false
    type = ConservativeAdvection
    variable = u
    velocity = '1 0 0'
  [../]
[]

[DGKernels]
  [./dg_advection_u]
    implicit = false
    type = DGConvection
    variable = u
    velocity = '1 0 0'
  [../]
[]
```
[...]
_(from test/tests/dgkernels/1d_advection_dg/1d_advection_dg.i)_

I stumbled across this as I was trying to modify these simple test cases and noticed the redundant definitions of velocity in this case. Is this a requirement or rather a workaround so the DG schemes fit in nicely with the CG main code of MOOSE?

2. I've looked into the source of some DG Kernels and got the impression that the code uses the same basis for trial and test function, so adhering to the Bubnov instead of Petrov Galerkin method. That would fit into the concept of the "main" code as declared [here](https://mooseframework.inl.gov/source/problems/FEProblemBase.html#shape_functions). Is that correct? If so, wouldn't it be more advantageous in terms of assembling the mass matrix to allow for separate function spaces or is this too complicated to implement or otherwise not feasible? Ern and Guermod<sup>1</sup> address this issue e.g. in Remark 5.4 p. 225 and in chap. 5.7.

3. When looking at the source code and examples, I couldn't find input file entries or info about the upwinding scheme used - which to my knowledge would be required at least for convection-type problems. How is this done in MOOSE? From FV I know some schemes in order to get reasonable tradeoff between diffusivity and stability, e.g. with WENO or TVD schemes (MinMod, SuperBee, etc.). If that's something to still implement into MOOSE, would it be possible to carry these over from the rDG app? I've seen [here](https://mooseframework.inl.gov/modules/rdg/index.html) that some TVD limiters have already been introduced into the rDG scheme. If so, I'd try to do that for myself and when that yields satisfactory results, share the code of course with everyone.

Lastly, do you guys think that further developing DG Kernels in MOOSE is even worthwhile in comparison with the classic CG-FEM for dealing with hyperbolic PDEs? Surely, CG-FEM (at least as I know it) is better suited for elliptic and parabolic problems, but does anyone know how severe the performance disadvantage for hyperbolic problems is in comparison to e.g. FV or rDG in this case? I'm asking since e.g. in your tutorial sessions (which helped me hugely set foot into the MOOSE ecosystem btw) you deal with advection problems in a fairly confident manner, so I got the impression that problems of such type would not be a big deal for MOOSE. Also, according to your website, even the [Navier Stokes app](https://mooseframework.inl.gov/modules/navier_stokes/index.html) relies on the CG-FEM method which got me wondering quite a bit.

I know that's a lot of questions, sorry for the long text. Hopefully someone can help shed some light on these topics.

Thanks and have a good weekend,
Patrick

> <sup>1</sup>: A. Ern, J.-L. Guermond. Theory and Practice of Finite Elements, Springer 2004.

**Comment by aeslaughter** [comment]
@lindsayad You should probably help out here.

**Comment by lindsayad** [comment, marked as answer]
>     1. First of all, I'm wondering why you need to specify the DG Kernels separately as an addition to a regular kernel, instead of replacing it - see e.g.:

We use Beatrice Riviere's "Discontinuous Galerkin Methods for Solving Elliptic and Parabolic Equations: Theory and Implementation" text for implementing the relatively rudimentary objects you are looking at. The object naming is definitely confusing. One would very reasonably think that `DGConvection` does everything needed for a DG method...but it doesn't. It instead implements only the convection interior face term. So If you have Riviere's text, `DGConvection` implements the middle term of equation 4.9. You still need `ConservativeAdvection` to capture the first term of equation 4.9, e.g. the elemental contributions.

>     2. I've looked into the source of some DG Kernels and got the impression that the code uses the same basis for trial and test function, so adhering to the Bubnov instead of Petrov Galerkin method. That would fit into the concept of the "main" code as declared [here](https://mooseframework.inl.gov/source/problems/FEProblemBase.html#shape_functions). Is that correct? If so, wouldn't it be more advantageous in terms of assembling the mass matrix to allow for separate function spaces or is this too complicated to implement or otherwise not feasible? Ern and Guermod1 address this issue e.g. in Remark 5.4 p. 225 and in chap. 5.7.

This is definitely true for the objects you are looking at. However, an application developer/MOOSE user has complete flexibility to choose the finite element method they wish. An example of this from continuous finite elements draws from the `ADKernelStabilized` and its child `ADKernelSUPG` where the user "strong residual" (which should look exactly like terms in the non-integrated/strong PDE) is multiplied by velocity * tau * grad_test. Here `grad_test` is indeed based off the same finite element family and order as the shape function used to evaluate the approximate solution, but this is still a piece of a Petrov-Galerkin method. In general you can write your own kernels, and define `computeQpResidual` (or its relatives) however you wish. E.g. you are not constrained to multiply by the `_test` member that corresponds to the Bubnov-Galerkin method.

>     3\. When looking at the source code and examples, I couldn't find input file entries or info about the upwinding scheme used - which to my knowledge would be required at least for convection-type problems. How is this done in MOOSE? From FV I know some schemes in order to get reasonable tradeoff between diffusivity and stability, e.g. with WENO or TVD schemes (MinMod, SuperBee, etc.). If that's something to still implement into MOOSE, would it be possible to carry these over from the rDG app? I've seen [here](https://mooseframework.inl.gov/modules/rdg/index.html) that some TVD limiters have already been introduced into the rDG scheme. If so, I'd try to do that for myself and when that yields satisfactory results, share the code of course with everyone.

`DGConvection` uses simple upwinding. We would certainly welcome more advanced convection discretizations in the framework or in our modules. To be honest, the DG kernels that are currently in the framework are probably not used by anyone in "production"-level simulations. I do not know too much about our rDG module. Perhaps @andrsd or @joshuahansel can comment more on that. I believe most of rDG development to date has been done with 1D target applications, but I'm not going to say that it won't work in multi-dimensions (simply because I do not know). We have been focusing a lot of development recently on finite volume in MOOSE and we have worked mostly on HLLC for free-flow and porous versions of the Euler equations. I know that rDG has some code for WENO/TVD schemes, including the MinMod and SuperBee that you mentioned.


> Lastly, do you guys think that further developing DG Kernels in MOOSE is even worthwhile in comparison with the classic CG-FEM for dealing with hyperbolic PDEs? Surely, CG-FEM (at least as I know it) is better suited for elliptic and parabolic problems, but does anyone know how severe the performance disadvantage for hyperbolic problems is in comparison to e.g. FV or rDG in this case? I'm asking since e.g. in your tutorial sessions (which helped me hugely set foot into the MOOSE ecosystem btw) you deal with advection problems in a fairly confident manner, so I got the impression that problems of such type would not be a big deal for MOOSE. Also, according to your website, even the [Navier Stokes app](https://mooseframework.inl.gov/modules/navier_stokes/index.html) relies on the CG-FEM method which got me wondering quite a bit.

I do think that DG/rDG/FV development/use in advection dominated applications is critical/advantageous when compared to CG-FEM. We have done a fair amount of simulation of incompressible Navier-Stokes with CFEM, but I think that in the long haul our discontinuos schemes are going to win out and become the schemes predominantly used by our user-base. CFEM in general struggles around any area of discontinuity whether it be at a corner where you have different boundary conditions meeting, or near shocks. Here is a very relevant figure of CFEM failing to handle the canonical Sod shock-tube problem:
![1D_shock_tube_10sec](https://user-images.githubusercontent.com/10248304/112217010-6a296e00-8bdf-11eb-94b8-568a1d4746d2.png)

Finite volume development is very heavy right now. I'll point you to the recently merged https://github.com/idaholab/moose/pull/15644 and open pull request https://github.com/idaholab/moose/pull/17088. I also have a CNSFV branch that I'm currently working on that we'll be submitting a PR to MOOSE for once we receive approval from export control. And with respect to more sophisticated DG/rDG objects compared to those currently in the framework I'll leave that to @andrsd and @joshuahansel

**Reply by lindsayad** [reply]
@pzimbrod if you have any follow-on questions, please let me know. We certainly welcome new contributors!!

**Comment by joshuahansel** [comment]
> When looking at the source code and examples, I couldn't find input file entries or info about the upwinding scheme used

The rDG module has only implemented the scalar advection equation. Its upwinding scheme is implemented in `AEFVUpwindInternalSideFlux`, and this is the general strategy - you implement your numerical flux scheme in a flux user object.

> From FV I know some schemes in order to get reasonable tradeoff between diffusivity and stability, e.g. with WENO or TVD schemes (MinMod, SuperBee, etc.). If that's something to still implement into MOOSE, would it be possible to carry these over from the rDG app?

There is no generic WENO/TVD capability in the rDG module, just the implementation for the scalar advection equation. The rDG mostly just provides base classes that can be useful for implementing an rDG scheme.

**Comment by pzimbrod** [comment]
Thank you both for the valuable insights! Definitely helped me a lot wrapping my head around the DG functionality that's already implemented in MOOSE. I'll look into that subject a little more in the next time and make sure to submit a PR as soon as I get a fruitful solution out of these ideas. I won't dare to make any promises since I come from using OpenFOAM and am rather new to the MOOSE style of getting things done, but definitely will give it a shot.

Please keep up the good work, I'd love to see projects like these pick up momentum in all types of applications!
